File: scala/Ledger.scala
package tqlmate

import com.typedb.driver.api.Driver
import com.typedb.driver.api.Transaction
import com.typedb.driver.api.concept.Concept

import scala.jdk.CollectionConverters.*
import scala.jdk.OptionConverters.*
import scala.util.Using

object Ledger:
  val Entity = "_tqlmate_schema_migration"
  val AttrVersion = "_tqlmate_version"
  val AttrAppliedAt = "_tqlmate_applied_at"

  private val EnsureSchema =
    s"""define
       |  attribute $AttrVersion, value string;
       |  attribute $AttrAppliedAt, value datetime;
       |  entity $Entity,
       |    owns $AttrVersion @card(1),
       |    owns $AttrAppliedAt @card(1);
       |""".stripMargin

  def ensure(driver: Driver, database: String): Result[Unit] =
    appliedVersions(driver, database) match
      case Right(_) => Right(())
      case Left(_)  => schemaQueries(driver, database, List(EnsureSchema))

  def appliedVersions(driver: Driver, database: String): Result[List[Version]] =
    edge:
      Using.resource(driver.transaction(database, Transaction.Type.READ)): tx =>
        val answer = tx.query(s"match $$m isa $Entity, has $AttrVersion $$v;").resolve()
        val versions =
          answer
            .asConceptRows()
            .asScala
            .toList
            .flatMap: row =>
              row.get("v").asScala.toList.flatMap: concept =>
                stringValue(concept).map(Version(_))
        versions.sorted

  def recordInsert(version: Version, appliedAt: String): String =
    s"""insert $$_ isa $Entity, has $AttrVersion "${version.value}", has $AttrAppliedAt $appliedAt;"""

  def recordDelete(version: Version): String =
    s"""match $$m isa $Entity, has $AttrVersion "${version.value}"; delete $$m;"""

  def schemaQueries(driver: Driver, database: String, queries: List[String]): Result[Unit] =
    edge:
      Using.resource(driver.transaction(database, Transaction.Type.SCHEMA)): tx =>
        queries.foreach: q =>
          val trimmed = q.trim
          if trimmed.nonEmpty then
            tx.query(trimmed).resolve()
        tx.commit()

  private def stringValue(concept: Concept): Option[String] =
    val opt = concept.tryGetString()
    if opt.isPresent then Some(opt.get) else None

  private def edge[A](body: => A): Result[A] =
    try Right(body)
    catch case e: Exception => Left(Error.fromThrowable(e))

File: src/ledger.rs
use futures::StreamExt;
use typedb_driver::{TransactionType, TypeDBDriver};

use crate::migration::{MigrationFile, Version};
use crate::{Error, Result};

pub const ENTITY: &str = "_tqlmate_schema_migration";
pub const ATTR_VERSION: &str = "_tqlmate_version";
pub const ATTR_APPLIED_AT: &str = "_tqlmate_applied_at";

const ENSURE_SCHEMA: &str = concat!(
    "define\n",
    "  attribute _tqlmate_version, value string;\n",
    "  attribute _tqlmate_applied_at, value datetime;\n",
    "  entity _tqlmate_schema_migration,\n",
    "    owns _tqlmate_version @card(1),\n",
    "    owns _tqlmate_applied_at @card(1);\n",
);

/// Bootstrap TypeQL that creates the `_tqlmate_*` ledger types (pure string; no network).
pub fn bootstrap_schema() -> &'static str {
    ENSURE_SCHEMA
}

pub async fn ensure(driver: &TypeDBDriver, database: &str) -> Result<()> {
    if applied_versions(driver, database).await.is_ok() {
        return Ok(());
    }
    schema_queries(driver, database, &[bootstrap_schema()]).await
}

pub async fn applied_versions(driver: &TypeDBDriver, database: &str) -> Result<Vec<Version>> {
    let tx = driver.transaction(database, TransactionType::Read).await?;
    let query = format!("match $m isa {ENTITY}, has {ATTR_VERSION} $v;");
    let answer = match tx.query(query).await {
        Ok(a) => a,
        Err(e) => {
            let _ = tx.close().await;
            return Err(e.into());
        }
    };
    let mut versions = Vec::new();
    let mut rows = answer.into_rows();
    while let Some(row) = rows.next().await {
        let row = row?;
        let concept = row
            .get("v")?
            .ok_or_else(|| Error::msg("missing version column"))?;
        let version = concept
            .try_get_string()
            .ok_or_else(|| Error::msg("version is not a string"))?
            .to_string();
        versions.push(Version::new(version));
    }
    let _ = tx.close().await;
    versions.sort();
    Ok(versions)
}

pub fn record_insert(version: &Version, applied_at: &str) -> String {
    format!(
        "insert $_ isa {ENTITY}, has {ATTR_VERSION} \"{}\", has {ATTR_APPLIED_AT} {applied_at};",
        version.as_str()
    )
}

pub fn record_delete(version: &Version) -> String {
    format!(
        "match $m isa {ENTITY}, has {ATTR_VERSION} \"{}\"; delete $m;",
        version.as_str()
    )
}

/// Header comments written by `dump`, listing applied migration labels.
pub fn dump_header(applied: &[Version], files: &[MigrationFile]) -> String {
    let mut out = String::from("-- Schema dumped by tqlmate\n-- Applied migrations:\n");
    if applied.is_empty() {
        out.push_str("--   (none)\n");
    } else {
        for v in applied {
            match files.iter().find(|f| &f.version == v) {
                Some(m) => out.push_str(&format!("--   {}\n", m.label())),
                None => out.push_str(&format!("--   {v}\n")),
            }
        }
    }
    out.push('\n');
    out
}

pub fn strip_dump_header(text: &str) -> String {
    text.lines()
        .skip_while(|l| {
            let t = l.trim();
            t.is_empty() || t.starts_with("--")
        })
        .collect::<Vec<_>>()
        .join("\n")
}

pub async fn schema_queries(driver: &TypeDBDriver, database: &str, queries: &[&str]) -> Result<()> {
    let tx = driver
        .transaction(database, TransactionType::Schema)
        .await?;
    for q in queries {
        let trimmed = q.trim();
        if trimmed.is_empty() {
            continue;
        }
        if let Err(e) = tx.query(trimmed).await {
            let _ = tx.close().await;
            return Err(e.into());
        }
    }
    tx.commit().await?;
    Ok(())
}
